### ISSUE TITLE
"Fetch type out of range" error when calling result.prior(), result.first(), result.last() or result.at_end() 

### ISSUE THREAD

**Issue by thibaudk** [nanodbc/nanodbc#368]
## Environment

* nanodbc version: 2.13.0-3 installed from [aur](https://aur.archlinux.org/packages/nanodbc)
* DBMS name/version: ```Microsoft SQL Server 2019 (RTM-GDR) (KB5014356) - 15.0.2095.3 (X64) Apr 29 2022 18:00:13 Copyright (C) 2019 Microsoft Corporation Express Edition (64-bit) on Windows Server 2016 Standard 10.0 <X64> (Build 14393: ) (Hypervisor)```
* ODBC connection string: i simply initialize the the connection object with {"dbName", "userName", "pwd"}
* OS and Compiler: Archlinux, clang version 14.0.6
* CMake settings: only -Dnanodbc_DIR:PATH=/usr/lib/cmake/nanodbc is set

## Actual behavior

calling result.prior(), result.first(), result.last() or result.at_end() throws the following error:
```nanodbc-2.13.0/nanodbc/nanodbc.cpp:2789: HY10: [Microsoft][ODBC Driver 18 for SQL Server]Fetch type out of range```

I also noticed that result.position() always returns 0.
Basically, only result.next() seem to affect the current result set.

## Disclaimer

I may be simply missing something in the configuration of unixODBC, or it could be a limitation of connecting to mssql on a windows server form a Linux machine. My apologies if this is not an actual issue of nanodbc.

**Comment by mloskot** [comment]
Searching https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlfetchscroll-function for `fetch type out of range` yields:
![image](https://user-images.githubusercontent.com/80741/210248649-5a9895dc-d8e2-46b3-bcb8-8ba5360e8f24.png)


Looks like we're back to the square of configurable cursors
https://github.com/nanodbc/nanodbc/issues/93#issuecomment-358245887

**Comment by lexicalunit** [comment]
This is a cursor configuration matter, and there is a working answer — with a caveat about which C++ standard you build against.

Reproduced your error exactly, then fixed it by asking for a scrollable cursor:

```
default cursor   : next=1 first=threw: HY106 [Microsoft][ODBC Driver 18 for SQL Server]Fetch type out of range
SQL_CURSOR_STATIC: next=1 first=1 last=1 val=5 prior=1 val=4   scrollable
```

nanodbc leaves `SQL_ATTR_CURSOR_TYPE` at the ODBC default, which is `SQL_CURSOR_FORWARD_ONLY`. Under that, `SQLFetchScroll` legitimately rejects every fetch orientation except `SQL_FETCH_NEXT`, so `first()`, `last()`, `prior()` and `at_end()` cannot work and `position()` stays 0. That is the driver following the spec rather than a nanodbc bug — and it is the same conclusion as #93, as noted above.

The recipe:

```cpp
#include <sqlext.h>

std::list<nanodbc::statement::attribute> attrs{
    nanodbc::statement::attribute(
        SQL_ATTR_CURSOR_TYPE, 0, static_cast<std::uintptr_t>(SQL_CURSOR_STATIC))};

nanodbc::statement st(conn, attrs);
st.prepare(NANODBC_TEXT("select i from t order by i;"));
auto results = st.execute();
```

**The caveat.** `statement::attribute` is public only when nanodbc is built at C++17 or later, where `NANODBC_HAS_STD_VARIANT` is defined. At C++14, which is still the project default, that nested class is `private`, so the `statement(connection&, std::list<attribute> const&)` constructor exists in the header but cannot be called. I have filed that separately as a defect in its own right.

So: at C++17 this works today and I would call your issue answered. At C++14 there is no public way to reach it, which is worth fixing regardless of this issue.

Leaving open until the C++14 gap is closed, since that is what stands between this and "just works".
